### PR TITLE
 [exporter/loadbalancing] Fix inconsistent handling of endpoints without port

### PR DESCRIPTION
**Pull request by jvilhuber**
**Description:**
Fixes the problem where a dns-resolver is defined without a `Port` element. This SHOULD result in the default port (4317) being used, but leads to some problems. See linked issue.

The gist of the fix is to ensure that the strings in the lb.exporters map always have the port, and are always looked up with the port. Some of the unit-tests (logs and traces) can no longer happen as written and have been removed.

A deeper fix would likely require more refactoring, since the host (with or without port) is also used in the `items` ring.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/10321

**Testing:**
Unit tests performed.
Ran test-config from the linked issue, which worked.

**Documentation:**
No documentation added or changed.

### CODE REVIEW

**Review by jpkrohling — COMMENTED** [review]
This makes sense, looks like we had inconsistent values in some places. Have you experimented going the other direction? Removing the port from the existing place(s), adding it only on the backing exporter instantiation? Here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3fd860446371071d409abfc5b101ecc047df9b37/exporter/loadbalancingexporter/trace_exporter.go#L73

File: exporter/loadbalancingexporter/loadbalancer_test.go [hunk 1]
```diff
@@ -190,20 +190,22 @@ func TestOnBackendChanges(t *testing.T) {
...
+	p.addMissingExporters(context.Background(), []string{"endpoint-1", "endpoint-2"})
 	resolved := []string{"endpoint-1"}
 
 	// test
 	p.removeExtraExporters(context.Background(), resolved)
 
 	// verify
 	assert.Len(t, p.exporters, 1)
-	assert.NotContains(t, p.exporters, "endpoint-2")
+	assert.Contains(t, p.exporters, endpointWithPort("endpoint-1"))
```
[the remark this answers is not in the document]
Comment (jvilhuber) [reply exporter/loadbalancingexporter/loadbalancer_test.go:208]: Sure. I figured asserting that the Len is 1, and that the 1 is endpoint-1 should cover all cases. But I can revert that.

File: exporter/loadbalancingexporter/log_exporter_test.go [hunk 2]
```diff
@@ -160,40 +160,6 @@ func TestConsumeLogs(t *testing.T) {
 	assert.Nil(t, res)
 }
 
-func TestConsumeLogsExporterNotFound(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (jvilhuber) [reply exporter/loadbalancingexporter/log_exporter_test.go:163]: It no longer works. It seemed to rely on the bug I fixed, i.e. that removeMissing would remove some endpoints. There's no way to make 
``` 	res := p.ConsumeLogs(context.Background(), simpleLogs())```
fails in this way.

File: exporter/loadbalancingexporter/trace_exporter_test.go [hunk 3]
```diff
@@ -227,40 +227,6 @@ func TestServiceBasedRoutingForSameTraceId(t *testing.T) {
 	}
 }
 
-func TestConsumeTracesExporterNotFound(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (jvilhuber) [reply exporter/loadbalancingexporter/trace_exporter_test.go:230]: Ditto. It relied on the broken behavior to 'work'.

**Comment by jvilhuber** [comment]
> This makes sense, looks like we had inconsistent values in some places. Have you experimented going the other direction? Removing the port from the existing place(s), adding it only on the backing exporter instantiation?

No, I didn't consider that, mainly because I think that could cause issues if we wind up with an endpoint with the same hostname, but different ports, like `endpoint-1:4317` and `endpoint-1:4318` or something. Not sure that's realistic, but it could be configured. I can have a look, though.

**Comment by jpkrohling** [comment]
> but with the static resolver? Any future resolver?

True, it's indeed realistic to have that as part of the static resolver. I think we even use that in the examples.

**Review by jpkrohling — APPROVED** [review]

**Comment by jvilhuber** [comment]
Fixes #10321

**Pull request merged**